target_chains/starknet/contracts/src/wormhole.cairo

```diff
@@ -1,29 +1,322 @@
+
 #[starknet::interface]
-trait IWormhole<T> {
-    fn get_value(self: @T) -> felt252;
-    fn set_value(ref self: T, name: felt252);
+pub trait IWormhole<T> {
+    fn submit_new_guardian_set(ref self: T, set_index: u32, guardians: Array<felt252>);
+    fn parse_and_verify_vm(ref self: T, encoded_vm: ByteArray) -> Result<VM, felt252>;
+}
+
+#[derive(Drop, Debug, Clone, Serde)]
+pub struct GuardianSignature {
+    pub guardian_index: u8,
+    pub signature: Signature,
+}
+
+#[derive(Drop, Debug, Clone, Serde)]
+pub struct VM {
+    pub version: u8,
+    pub guardian_set_index: u32,
+    pub signatures: Array<GuardianSignature>,
+    pub timestamp: u32,
+    pub nonce: u32,
+    pub emitter_chain_id: u16,
+    pub emitter_address: u256,
+    pub sequence: u64,
+    pub consistency_level: u8,
+    pub payload: ByteArray,
+}
+
+pub mod error_codes {
+    pub const NO_GUARDIANS_SPECIFIED: felt252 = 'no guardians specified';
+    pub const TOO_MANY_GUARDIANS: felt252 = 'too many guardians';
+    pub const INVALID_GUARDIAN_KEY: felt252 = 'invalid guardian key';
+    // guardian set index must increase in steps of 1
+    pub const INVALID_GUARDIAN_SET_SEQUENCE: felt252 = 'invalid guardian set sequence';
+    pub const ACCESS_DENIED: felt252 = 'access denied';
+
+    pub const VM_VERSION_INCOMPATIBLE: felt252 = 'VM version incompatible';
+    pub const INVALID_GUARDIAN_SET_INDEX: felt252 = 'invalid guardian set index';
+    pub const INVALID_SIGNATURE: felt252 = 'invalid signature';
+    pub const GUARDIAN_SET_EXPIRED: felt252 = 'guardian set expired';
+    pub const NO_QUORUM: felt252 = 'no quorum';
+    pub const INVALID_SIGNATURE_ORDER: felt252 = 'invalid signature order';
+    pub const INVALID_GUARDIAN_INDEX: felt252 = 'invalid guardian index';
+}
+
+pub fn quorum(num_guardians: usize) -> usize {
+    assert(num_guardians < 256, error_codes::TOO_MANY_GUARDIANS);
+    ((num_guardians * 2) / 3) + 1
 }
 
 #[starknet::contract]
 mod wormhole {
+    use core::box::BoxTrait;
+    use core::array::ArrayTrait;
+    use super::{VM, IWormhole, GuardianSignature, error_codes, quorum};
+    use pyth::reader::{Reader, ReaderImpl, ByteArray, UNEXPECTED_OVERFLOW};
+    use core::starknet::secp256_trait::{Signature, recover_public_key, Secp256PointTrait};
+    use core::starknet::secp256k1::Secp256k1Point;
+    use core::starknet::{
+        ContractAddress, get_execution_info, get_caller_address, get_block_timestamp
+    };
+    use core::keccak::cairo_keccak;
+    use core::integer::u128_byte_reverse;
+    use core::panic_with_felt252;
+
+    #[derive(Drop, Debug, Clone, Serde, starknet::Store)]
+    struct GuardianSet {
+        num_guardians: usize,
+        // XXX: storage doesn't work if we use Option here.
+        expiration_time: u64,
+    }
+
     #[storage]
     struct Storage {
-        name: felt252,
+        owner: ContractAddress,
+        current_guardian_set_index: u32,
+        guardian_sets: LegacyMap<u32, GuardianSet>,
```

Comment:
Can you write a comment on why LegacyMap is used? (I don't have the context but the name has Legacy in it and it brings this question of why). I wonder whether we could use Felt256Dict here.

Reply:
My understanding is that in theory there are Legacy (pedersen based) and NonLegacy (poseidon based) storage mapping formats, but NonLegacy mapping is not yet implemented, and there isn't any "non-legacy" map available. The book [just says to use LegacyMap](https://book.cairo-lang.org/ch14-01-contract-storage.html) without any explanation. `Felt252Dict` isn't usable here because it doesn't implement `starknet::Store`. So there is no alternative to `LegacyMap` right now.